### PR TITLE
produce less garbage while generating moves

### PR DESCRIPTION
**Pull request by niklasf**
Based on an optimization idea by https://lichess.org/@/Vetinari_Computer. Needs more benchmarking and testing.

### CODE REVIEW

File: src/main/java/game/Board.java [hunk 1]
```diff
@@ -254,7 +253,7 @@ private long attacksTo(int sq, boolean attacker, long occupied) {
             Bitboard.pawnAttacks(!attacker, sq) & this.pawns);
     }
 
-    public void legalMoves(ArrayList<Move> moves) {
+    public void legalMoves(MoveList moves) {
```
Comment (ddugovic) [inline src/main/java/game/Board.java:256]: Perhaps further performance improvement (of this code) is possible through usage of the Java 8 Stream API?

File: src/main/java/game/PerftTest.java [hunk 2]
```diff
@@ -1,20 +1,19 @@
...
 
-import java.util.ArrayList;
-
 public class PerftTest {
     public static long perft(Board board, int depth) {
         // Counts the number of paths in the tree of legal moves. Useful for
         // testing performance and comparing to well known numbers.
         if (depth >= 1) {
-            ArrayList<Move> moves = new ArrayList<Move>(255);
+            MoveList moves = new MoveList();
```
Comment (ddugovic) [inline src/main/java/game/PerftTest.java:8]: Given the importance of ordering elements, I wonder if `MoveList` should extend `TreeSet<Move>` or if that would be overkill.

File: src/main/java/game/MoveList.java [hunk 3]
```diff
@@ -56,18 +59,31 @@ public void sort() {
...
-                i += 1;
+                i++;
             } else {
                 swapRemove(i);
             }
         }
     }
 
     private void swapRemove(int i) {
+        // Removes a move by swapping the last move into its place.
```
Comment (ddugovic) [inline src/main/java/game/MoveList.java]: This first line could be put in javadoc style:

```
     /**
      * Removes a move by swapping the last move into its place
      */
     private void swapRemove(int i) {
```

Comment (ornicar) [reply src/main/java/game/MoveList.java]: Python hacker spotted.

**Pull request merged**